src/appleseed/renderer/meta/tests/test_phasefunction.cpp

```diff
@@ -0,0 +1,337 @@
+            m_base_parameters
+                .insert("scattering", "0.5")
+                .insert("scattering_multiplier", "1.0")
+                .insert("extinction", "0.5")
+                .insert("extinction_coefficient", "1.0");
+        }
+
+        template <typename ReturnType, typename Procedure>
+        ReturnType setup_environment_and_evaluate(Procedure procedure)
+        {
+            TextureStore texture_store(m_scene);
+            TextureCache texture_cache(texture_store);
+
+            boost::shared_ptr<OIIO::TextureSystem> texture_system(
+                OIIO::TextureSystem::create(),
+                boost::bind(&OIIO::TextureSystem::destroy, _1));
+
+            RendererServices renderer_services(
+                m_project,
+                *texture_system);
+
+            boost::shared_ptr<OSL::ShadingSystem> shading_system(
+                new OSL::ShadingSystem(&renderer_services, texture_system.get()));
+
+            Intersector intersector(
+                m_project.get_trace_context(),
+                texture_cache);
+
+            Arena arena;
+            OSLShaderGroupExec sg_exec(*shading_system, arena);
+
+            Tracer tracer(
+                m_scene,
+                intersector,
+                texture_cache,
+                sg_exec);
+
+            ShadingContext shading_context(
+                intersector,
+                tracer,
+                texture_cache,
+                *texture_system,
+                sg_exec,
+                arena,
+                0);
+
+            return procedure(shading_context, arena);
+        }
+
+        // Integrate PDF of phase function (direction sampling) using straightforward Monte-Carlo approach.
+        static float integrate_phase_function_direction_pdf(
+            PhaseFunction& phase_function,
+            ShadingContext& shading_context,
+            Arena& arena)
+        {
+            ShadingRay shading_ray;
+            shading_ray.m_org = Vector3d(0.0f, 0.0f, 0.0f);
+            shading_ray.m_dir = Vector3d(1.0f, 0.0f, 0.0f);
+            void* data = phase_function.evaluate_inputs(shading_context, shading_ray);
+            phase_function.prepare_inputs(arena, shading_ray, data);
+
+            MersenneTwister rng;
+            SamplingContext sampling_context(rng, SamplingContext::RNGMode);
+
+            float integral = 0.0f;
+            for (int i = 0; i < NumberOfSamples; ++i)
+            {
+                sampling_context.split_in_place(2, 1);
+                Vector2f s = sampling_context.next2<Vector2f>();
+                Vector3f incoming = sample_sphere_uniform<float>(s);
+                integral += phase_function.evaluate(shading_ray, data, 0.5f, incoming);
+            }
+
+            return integral * foundation::FourPi<float>() / NumberOfSamples;
+        }
+
+        // Check if probabilistic sampling is consistent with the returned PDF values.
+        static Vector3f check_direction_sampling_consistency(
```

Comment:
Since this method doesn't check anything, I suggest renaming it. Unfortunately I don't have a good suggestion right now since I'm not 100% sure what this does :) We expect the result to be a zero vector, correct?